File: 1208.尽可能使字符串相等/solution.go
func equalSubstring(s string, t string, maxCost int) int {
	sz := len(s)
	cst := make([]int, sz)
	for i := 0; i < sz; i++ {
		cst[i] = abs(int(s[i]) - int(t[i]))
	}
	var ans int
	var l, r int
	var sum int
	for ; l < sz; l++ {
		for ; r < sz && sum+cst[r] <= maxCost; r++ {
			sum += cst[r]
		}
		ans = max(ans, r-l)
		sum -= cst[l]
	}
	return ans
}

func abs(n int) int {
	if n < 0 {
		return -n
	}
	return n
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

File: 1208.尽可能使字符串相等/solution.cpp
// 滑动窗口
class Solution {
public:
    int equalSubstring(string s, string t, int maxCost) {
        int sz=s.size();
        vector<int> cst(sz);
        for (int i=0;i<sz;i++) {
            cst.at(i)=abs((int)s.at(i)-(int)t.at(i));
        }
        int ans=0;
        int r=0,l=0;
        int sum=0;
        for (l=0;l<sz;l++) {
            while (r<sz && sum+cst.at(r)<=maxCost) {
                sum+=cst.at(r);
                r++;
            }
            ans=max(r-l,ans);
            sum-=cst.at(l);
        }
        return ans;
    }
};
